### DISCUSSION TITLE
Third-party apps integration

### DISCUSSION THREAD

**Discussion by kevinpapst** [kimai/kimai#684] [General]
I want to collect ideas about third-party integrations.

There are many SaaS software around project-management, which could be coupled with Kimai:
- for synching customer/projects/tasks
- for automatic time-tracking
- for import/export

If you submit an idea, please include your use-case and what should be possible. If you can supply that, also leave some links here to their API.

**Comment by kevinpapst** [comment]
Request via email...
Asana: https://asana.com/developers/documentation/examples-tutorials/overview

**Comment by kevinpapst** [comment]
Dollibarr, see #471

**Comment by j0hannesr0th** [comment]
Not a project management software but a SaaS: Lexoffice - https://www.lexoffice.de/funktionen/api/

**Reply by kevinpapst** [reply]
Lexoffice has this [public API ](https://developers.lexoffice.io/docs/#lexoffice-api-documentation/)

**Comment by simone-gasparini** [comment]
Atlassian Jira to track issue/project time

**Comment by infeeeee** [comment]
[Redmine](https://www.redmine.org/) open source project management platform has a rest api where you can add time entries to issues: https://www.redmine.org/projects/redmine/wiki/Rest_TimeEntries. Also projects and users can be created or read by the api.

**Reply by pkoevesdi** [reply]
I'd join that too, although I already 'wrote' an interface workflow in n8n for one-directional syncing of redmine projects, issues, activities and bi-directional syncing of time-entries. Willing to share, if somebody is interested.

**Reply by stehajo** [reply]
Oh yes, that would be awesome! Please share your sync-tool, it could give me a kickstart!

**Reply by pkoevesdi** [reply]
Ok, here it is, with only few documentation and without support or warranty:

1. Since Kimai sadly doesn't support any Webhooks yet (see https://github.com/kimai/kimai/issues/1407), we need some means to recognize the change in the kimai database. I do that with the BinLogStreamReader from https://pypi.org/project/mysql-replication/, implemented in this python script:
```
# This script watches the mySQL database for changes (in it's binary log) and throws the events to a REST API

import json
import base64
from datetime import date, datetime
import requests

from pymysqlreplication import BinLogStreamReader
from pymysqlreplication.row_event import (
    DeleteRowsEvent, UpdateRowsEvent, WriteRowsEvent)

WATCHER_SETTINGS = {
    "host": "127.0.0.1",
    "user": "binlogwatcher",
    "passwd": "password"
}

REQUEST_SETTINGS = {
    "URL": "https://automate.example.com/webhook/kimai/",

}


class JsonCustEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, bytes):
            return base64.b64encode(o).decode("ascii")
        elif isinstance(o, (datetime, date)):
            return o.isoformat()
        else:
            return super().default(o)


def main():

    stream = BinLogStreamReader(connection_settings=WATCHER_SETTINGS,
                                server_id=1,
                                only_events=[DeleteRowsEvent,
                                             WriteRowsEvent, UpdateRowsEvent],
                                only_tables=['kimai2_absence',
                                             'kimai2_timesheet'],
                                blocking=True)

    for binlogevent in stream:
        if not getattr(binlogevent, 'rows', None):
            continue
        rows = json.dumps(binlogevent.rows, cls=JsonCustEncoder)
        data = {'eventname': type(binlogevent).__name__,
                'rows': json.loads(rows)}
        r = requests.post(
            url=REQUEST_SETTINGS['URL']+binlogevent.table, json=data)
        print(f"Passed {data['eventname']} to Webhook, Result: {r.status_code}")

    stream.close()

if __name__ == "__main__":
    main()
```
running with that service:
```
# to install, execute `systemctl enable ./binlogwatcher.service; systemctl daemon-reload; systemctl start binlogwatcher.service`

[Unit]
Description=Binary Log Watcher for Kimai database
After=syslog.target

[Service]
Type=simple
WorkingDirectory=/home/user/binlogwatcher/
ExecStart=/home/user/binlogwatcher/.venv/bin/python -m binlogwatcher
Restart=on-failure
#StandardOutput=syslog
#StandardError=syslog

[Install]
WantedBy=multi-user.target
```
2. Redmine 5.0.4 supports lot of webhooks, but 2 were missing, So i added them, here's the patch:
```diff --git a/redmine/app/controllers/timelog_controller.rb b/redmine/app/controllers/timelog_controller.rb
index dece5c4..12a3e91 100644
--- a/redmine/app/controllers/timelog_controller.rb
+++ b/redmine/app/controllers/timelog_controller.rb
@@ -117,6 +117,8 @@ class TimelogController < ApplicationController
               {:params => params, :time_entry => @time_entry})
 
     if @time_entry.save
+      call_hook(:controller_timelog_edit_after_save,
+      {:params => params, :time_entry => @time_entry})
       respond_to do |format|
         format.html do
           flash[:notice] = l(:notice_successful_create)
@@ -232,6 +234,8 @@ class TimelogController < ApplicationController
   def destroy
     destroyed = TimeEntry.transaction do
       @time_entries.each do |t|
+        call_hook(:controller_timelog_edit_before_destroy,
+        {:params => params, :time_entry => t})        
         unless t.destroy && t.destroyed?
           raise ActiveRecord::Rollback
         end
```
3. then, in n8n I have these several seperate workflows, that do the magic, using a local mysql database (that has to be made before, see the workflows how it is to be set up, deduct which tables you need):
[Upsert Ticket in Kimai.json](https://github.com/user-attachments/files/23810685/Upsert.Ticket.in.Kimai.json)
[Upsert Project in Kimai.json](https://github.com/user-attachments/files/23810686/Upsert.Project.in.Kimai.json)
[Upsert Timesheet in Kimai.json](https://github.com/user-attachments/files/23810687/Upsert.Timesheet.in.Kimai.json)
[Upsert Timesheet in Redmine.json](https://github.com/user-attachments/files/23810689/Upsert.Timesheet.in.Redmine.json)
[Upsert Activity in Kimai.json](https://github.com/user-attachments/files/23810690/Upsert.Activity.in.Kimai.json)
Btw.: With the current n8n version, You don't need an external MySQL databasem You can use one provided by n8n. Didn't try it, just read it

**Comment by infeeeee** [comment]
[Gitea](https://gitea.io/en-us/) self hosted git server has built in time tracking per issues, but time entries can be created with the api as well: https://app.swaggerhub.com/apis/gitea/gitea-api/1.4.2#/issue/issueAddTime

**Comment by AndyHeinz** [comment]
Hi,
those integrations would be great:
- Punch in via MS Teams Bot
- Timeular:  https://developers.timeular.com/public-api/
Thanks!

**Comment by nightvisi0n** [comment]
Integration with [GitLab](https://about.gitlab.com/) would be great:  
- https://docs.gitlab.com/ce/workflow/time_tracking.html  
- https://docs.gitlab.com/ee/api/issues.html#add-spent-time-for-an-issue

**Comment by Dzordzu** [comment]
EDIT: I hope this isn't a necrobump
Integration with [taiga.io](https://taiga.io/) would be a great option. 
> Taiga is a project management platform for agile developers & designers and project managers
* [API](https://taigaio.github.io/taiga-doc/dist/api.html)

It would be great if there will be an import/sync option with the current tasks

**Comment by septianw** [comment]
Would be nice if we have IDE plugin to track our time.

**Comment by ebdavison** [comment]
Obsidian
https://obsidian.md/

**Reply by kevinpapst** [reply]
What about Obsidian? Can you please share your ideas about this possible integration. What should it achieve and in which way?

**Reply by ebdavison** [reply]
Oh, sure.  So with Obsidian, there is are plugins that allow various things.  There is currently a toggl time tracking integration/plugin which allows for time reports, graphs, lists, etc.  I would like to do something similar so that on a daily note or monthly note, I can see an overview of the time entries for a given time period or a list of time entries for the month or day.  

The toggle plugin is here: https://github.com/mcndt/obsidian-toggl-integration

Something similar would be good.  Something better ( ;) ) would be great!

**Reply by ebdavison** [reply]
Mostly this is a typescript code base and would use the standard kimai API but I don't know typescript / javascript so just putting it here.  Maybe someone (if not you) might want to jump in and help ...

**Comment by pkoevesdi** [comment]
Absence syncing with outlook (MS365) calendars.
Although I 'wrote' an workflow in n8n for one-directional syncing of absences. Willing to share, if somebody is interested.

**Reply by ebdavison** [reply]
I would be interested in the n8n workflow

**Reply by pkoevesdi** [reply]
You need some authentication mechanism with Office 365. I made one here:
https://community.n8n.io/t/authenticate-with-microsoft-without-callback/56625/10
Alternatively You can use my updated mySQL based version of it:
[OAuth2_Device_Code.json](https://github.com/user-attachments/files/17962550/OAuth2_Device_Code.json)
[OAuth2_Refresh_Token.json](https://github.com/user-attachments/files/17962551/OAuth2_Refresh_Token.json)

Then You can use this:
[Kimai_Absence.json](https://github.com/user-attachments/files/17962547/Kimai_Absence.json)

Disclaimer: It works on our side, but there's no extra documentation. You'll have to understand what I did from the program itself and adapt it to Your needs and environment.

**Comment by jhamidi** [comment]
I don't know how widespread this is, but integration with Quickbooks would be good, can be just one way, ie. Kimai -> Quickbooks.

**Reply by bexodim** [reply]
Even an export in the quickbooks format would be helpful!
[sample_invoice_import..csv](https://github.com/user-attachments/files/20600227/sample_invoice_import.csv)

**Comment by jhamidi** [comment]
Github Project (similar to Atlassian JIRA functionality)

**Comment by kansoit** [comment]
Yes, please! I'm completely on board with better integrations. Honestly, I'm so tired claiming hours manually every day. It’s a huge time sink.

We need something that lets AI handle the boring stuff. My vote is for implementing an MCP Server.

I would like to use an AI agent to talk directly to Kimai. Instead of clicking around, I could just tell my AI assistant, 'Hey, log an hour on the Q4 report project.'

Since Kimai already has APIs, adding an MCP layer would just be the final step to unlock serious, natural-language automation. It would be a massive quality-of-life improvement!

**Comment by adn77** [comment]
I don't know if it fits here, I created an integration for the popular Timeular (formerly ZEI:) hardware:
 https://github.com/adn77/timeular

(possibly related to https://github.com/kimai/kimai/issues/285)

**Reply by kevinpapst** [reply]
Please create a separate post in https://github.com/kimai/kimai/discussions/categories/plugins so it has a better chance to be found

**Reply by adn77** [reply]
https://github.com/kimai/kimai/discussions/5933
